components/brave_wallet_ui/common/hardware/ledgerjs/eth_ledger_bridge_keyring.ts

```diff
@@ -1,152 +1,151 @@
+    if (!result.success) {
+      return result
+    }
+    const data = await this.sendCommand<EthSignTransactionResponse>({
+      command: LedgerCommand.SignTransaction,
+      id: LedgerCommand.SignTransaction,
+      path: path,
+      rawTxHex: rawTxHex,
+      origin: window.origin
+    })
+    if (data === LedgerBridgeErrorCodes.BridgeNotReady ||
+        data === LedgerBridgeErrorCodes.CommandInProgress) {
+      return this.createErrorFromCode(data)
+    }
+    if (!data.payload.success) {
+      const ledgerError = data.payload as LedgerError
+      return { success: false, error: ledgerError.message, code: ledgerError.statusCode }
+    }
 
-  makeApp = async () => {
-    this.app = new Eth(await TransportWebHID.create())
+    const responsePayload = data.payload as EthSignTransactionResponsePayload
+    return { success: true, payload: { v: responsePayload.v, r: responsePayload.r, s: responsePayload.s } }
   }
 
-  unlock = async (): Promise<HardwareOperationResult> => {
-    if (this.isUnlocked()) {
-      return { success: true }
+  signPersonalMessage = async (path: string, message: string): Promise<SignHardwareOperationResult> => {
+    const result = await this.unlock()
+    if (!result.success) {
+      return result
     }
-
-    if (!this.app) {
-      await this.makeApp()
+    const messageHex = Buffer.from(message).toString('hex')
+    const data = await this.sendCommand<EthSignPersonalMessageResponse>({
+      command: LedgerCommand.SignPersonalMessage,
+      id: LedgerCommand.SignPersonalMessage,
+      path: path,
+      origin: window.origin,
+      messageHex: messageHex
+    })
+    if (data === LedgerBridgeErrorCodes.BridgeNotReady ||
+        data === LedgerBridgeErrorCodes.CommandInProgress) {
+      return this.createErrorFromCode(data)
     }
-    if (this.app && !this.deviceId) {
-      const eth: Eth = this.app
-      eth.transport.on('disconnect', this.onDisconnected)
-      const zeroPath = this.getPathForIndex(0, LedgerDerivationPaths.LedgerLive)
-      const address = (await eth.getAddress(zeroPath)).address
-      this.deviceId = await hardwareDeviceIdFromAddress(address)
+    if (!data.payload.success) {
+      const ledgerError = data.payload as LedgerError
+      return { success: false, error: ledgerError.message, code: ledgerError.statusCode }
     }
-    return { success: this.isUnlocked() }
-  }
 
-  signTransaction = async (path: string, rawTxHex: string): Promise<SignHardwareOperationResult> => {
-    try {
-      const unlocked = await this.unlock()
-      if (!unlocked.success || !this.app) {
-        return unlocked
-      }
-      const eth: Eth = this.app
-      const signed = await eth.signTransaction(path, rawTxHex)
-      return { success: true, payload: signed }
-    } catch (e) {
-      return { success: false, error: e.message, code: e.statusCode || e.id || e.name }
+    const responsePayload = data.payload as EthSignPersonalMessageResponsePayload
```

Comment:
nit should we be type validating this payload rather than casting it?

Reply:
We do some validation, but let me know if there's a better approach here

`data` will have the type `EthSignPersonalMessageResponse` -
```
export type EthSignPersonalMessageResponse = CommandMessage & {
  payload: EthSignPersonalMessageResponsePayload | LedgerError
}
```
so the `payload` is either `EthSignPersonalMessageResponsePayload` or `LedgerError`.  If the result is `EthSignPersonalMessageResponsePayload`, then `payload.success` will be true, and false if it's a `LedgerError`.

Before typecasting, we check the value of `data.payload.success` - perhaps this is the validation you're thinking of?
```
if (!data.payload.success) {
  const ledgerError = data.payload as LedgerError
  return { success: false, error: ledgerError, code: ledgerError.statusCode }
}
const responsePayload = data.payload as EthGetAccountResponsePayload
```

I bet there is a better way to do it - perhaps not having union types at all and instead returning a single type may contain an error, but if that's the way we go, then that might be better covered in https://github.com/brave/brave-browser/issues/24443.

Reply:
Yeah that makes sense what's going on then. I'm good with leaving it to that refactor as well for it it's non-blocking for security anyways